### PR TITLE
feat(log): Add logging of various account events

### PR DESCRIPTION
**Pull request by chilts**
Events added are:

* account.created
* account.verified
* account.login
* account.signed
* account.deleted

### CODE REVIEW

**Comment by chilts** [comment]
Easy enough to add/remove any more events. This is just a starting point, ready for conversation. :)

**Comment by rfk** [comment]
Thanks @chilts, some quick thoughts:

* `"op: event"` seems a little too generic to me, and "event" in general has a lot of meanings in this pipeline.  I wonder if we can use a more descriptive noun here like "activityEvent" or even "accountActivityEvent" to disambiguate.
* To sanity-check, do we capture logs at `INFO` level in production?
* It may be worth making a custom `log.activtyEvent()` function to make it easier to iterate on the details.

We should also ensure we capture as much of the activity's context as we can.  At a minimum this includes:

* user-agent string
* "service" param if there's one in the request payload or query params

The above two will let us recognise activity like "logged into sync from a mobile device" and "created an account to use hello" which will be important for driving cross-service metrics and promotions.

**Comment by chilts** [comment]
Thanks @rfk. All good suggestions, will do. I didn't realise we wanted to log the UserAgent.

**Comment by rfk** [comment]
>  I didn't realise we wanted to log the UserAgent.

Neither did I until I had this PR as a starting point to think through the flow ;-)

**Comment by vladikoff** [comment]
I'm not sure how this metrics is going to pipe to statsd but this is how we set the User Agent tags on the content server: https://github.com/mozilla/fxa-content-server/blob/master/server/lib/statsd-collector.js#L39

**Comment by chilts** [comment]
Cool, thanks @vladikoff.

**Comment by vladikoff** [comment]
However, if you are sending the metrics to Heka then I think Heka already parses it for you.

**Comment by chilts** [comment]
In which case, processing out of band is probably a better option (of course, depending on what we want from it will depend on how we process it).

**Comment by rfk** [comment]
@chilts do you want to target this for train-39? please set milestone appropriately

**Comment by chilts** [comment]
@rfk Updated to add those extra fields and tried to figure out a nice way of doing it. r?

Aside: Yay for milestones (with dates)!

happy-spiderman

I won't paste all these, but perhaps a new resource for @pdehaan : http://sophlylaughing.blogspot.co.nz/2013/11/happy-gifs.html

File: lib/log.js [hunk 1]
```diff
@@ -61,6 +61,20 @@ Lug.prototype.event = function (name, data) {
...
 
+Lug.prototype.activityEvent = function(event, uid, request) {
+  var info = {
+    event: event,
+    uid: uid,
+  }
+  if ( request.headers['user-agent'] ) {
+    info.userAgent = request.headers['user-agent']
+  }
+  if ( request.payload && request.payload.service ) {
```
Comment (rfk) [inline lib/log.js]: nit: the space after the paren here is inconsistent with the style of the surrounding code

**Comment by rfk** [comment]
@dannycoates can you please provide a second set of eyes on what we're doing here?  I feel like I'm too close to the nitty-gritty details to sanity-check how it fits in big-picture.

**Comment by dannycoates** [comment]
This looks like it overlaps a lot with the existing `log.event`. That one is intended for broadcasting events that other services can subscribe to, whereas `log.activityEvent` writes them to the app log, which I assume is also for downstream use but mainly logging and stats(?)

I don't like seeing lines like this:
```js
log.event('delete', { uid: uid + '@' + domain })
log.activityEvent('account.deleted', uid, request)
```

Can we combine them? I'm not exactly sure what use case `activityEvent` is targeting so I don't know what to suggest.

**Comment by rfk** [comment]
> I don't like seeing lines like this:

Yeah, it's definitely a smell.

> I assume is also for downstream use but mainly logging and stats

Broadly, yes.  These events form a much more detailed stream of user activity for internal metrics purposes (and possibly plugging in to select internal services like basket).  They log extra data that's not included in the existing `log.event` lines, e.g. service and user-agent.

IIUC the `log.event` events will eventually go away, supplanted by the notification-event stuff in the db server, and the new notification server.  So it may make sense to have them live side-by-side for a short while rather than try to combine them with this new stream.

**Comment by dannycoates** [comment]
> supplanted by the notification-event stuff in the db server

Yeah, I forgot that we put that directly in the db.

Two sources for the (mostly) same data still seems suspicious to me. It makes me think one of them should go away, the subset.

I don't want to block this though.

**Comment by rfk** [comment]
@dannycoates should I interpret that as "r+ with followups"?

**Comment by rfk** [comment]
I wonder if we only really need a subset of these for metrics purposes.  Perhaps just "created", "login", and "signed" would be sufficient?  I can't think of any query we'd want to do on 'verify' or 'deleted' events that we couldn't answer from public data in the notification stream, while "this user logged in on a mobile device" is very useful for our purposes.

**Comment by chilts** [comment]
Now just created, login and signed.

**Pull request merged**